baseplate/secrets/store.py

```diff
@@ -0,0 +1,253 @@
+"""Application integration with the secret fetcher daemon."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import base64
+import binascii
+import collections
+import datetime
+import json
+import logging
+
+from .. import config
+from ..context import ContextFactory
+
+
+ISO_FORMAT = "%Y-%m-%dT%H:%M:%S.%fZ"
+
+logger = logging.getLogger(__name__)
+
+
+class SecretNotFoundError(Exception):
+    """Raised when the requested secret is not in the local vault."""
+
+    def __init__(self, name):
+        super(SecretNotFoundError, self).__init__()
+        self.name = name
+
+    def __str__(self):  # pragma: nocover
+        return "secret not found: {!r}".format(self.name)
+
+
+class CorruptSecretError(Exception):
+    """Raised when the requested secret does not match the expected format."""
+
+    def __init__(self, path, message):
+        super(CorruptSecretError, self).__init__()
+        self.path = path
+        self.message = message
+
+    def __str__(self):  # pragma: nocover
+        return "{}: {}".format(self.path, self.message)
+
+
+_VersionedSecret = collections.namedtuple(
+    "VersionedSecret", "previous current next")
+
+
+class VersionedSecret(_VersionedSecret):
+    """A versioned secret.
+
+    Versioned secrets allow for seamless rotation of keys. When using the
+    secret to generate tokens (e.g. signing a message) always use the
+    ``current`` value. When validating tokens, check against all the versions
+    in ``all_versions``. This will allow keys to rotate smoothly even if not
+    done instantly across all users of the secret.
+
+    """
+
+    @property
+    def all_versions(self):
+        """Return an iterator over the available versions of this secret."""
+        yield self.current
+
+        if self.previous is not None:
+            yield self.previous
+
+        if self.next is not None:
+            yield self.next
+
+    @classmethod
+    def from_simple_secret(cls, value):
+        """Make a fake versioned secret from a single value.
+
+        This is a backwards compatibility shim for use with APIs that take
+        versioned secrets. Try to use proper versioned secrets fetched from the
+        secrets store instead.
+
+        """
+        return cls(
+            previous=None,
+            current=value,
+            next=None,
+        )
+
+
+def _decode_secret(path, encoding, value):
+    if encoding == "identity":
```

Comment:
Why is this called `identity`?  Is it used for something special that makes it tend to not be random bytes that need to be encoded?  Also, I've noticed it is the default if no encoding is specified, so it almost feels like just a default non-encoding?  I guess I'm just not sure what it's for.